src/connections.rs

```diff
@@ -332,16 +176,178 @@ impl std::fmt::Debug for SendStream {
+    mut quinn_incoming: quinn::Incoming,
+    connection_pool: ConnectionPool,
+    message_tx: UnboundedSender<(SocketAddr, Bytes)>,
+    connection_tx: UnboundedSender<SocketAddr>,
+    disconnection_tx: UnboundedSender<SocketAddr>,
+) {
+    let _ = tokio::spawn(async move {
+        loop {
+            match quinn_incoming.next().await {
+                Some(quinn_conn) => match quinn_conn.await {
+                    Ok(quinn::NewConnection {
+                        connection,
+                        uni_streams,
+                        bi_streams,
+                        ..
+                    }) => {
+                        let peer_address = connection.remote_address();
+                        let pool_handle = connection_pool.insert(peer_address, connection);
+                        connection_tx
+                            .send(peer_address)
+                            .map_err(|err| Error::MpscChannelSend(err.to_string()))?;
+                        listen_for_incoming_messages(
+                            uni_streams,
+                            bi_streams,
+                            pool_handle,
+                            message_tx.clone(),
+                            disconnection_tx.clone(),
+                        );
+                    }
+                    Err(err) => {
+                        error!("An incoming connection failed because of an error: {}", err);
+                    }
+                },
+                None => {
+                    trace!("quinn::Incoming::next() returned None. There will be no more incoming connections");
+                    break;
+                }
+            }
+        }
+        Ok::<_, Error>(())
+    });
+}
+
+pub(super) fn listen_for_incoming_messages(
+    mut uni_streams: quinn::IncomingUniStreams,
+    mut bi_streams: quinn::IncomingBiStreams,
+    remover: ConnectionRemover,
+    message_tx: UnboundedSender<(SocketAddr, Bytes)>,
+    disconnection_tx: UnboundedSender<SocketAddr>,
+) {
+    let src = *remover.remote_addr();
+    let _ = tokio::spawn(async move {
+        loop {
+            log::info!("Listening for new messages on new thread");
+            let message: Option<Message> = select! {
+                next_uni = next_on_uni_streams(&mut uni_streams) =>
+                next_uni.map(|(bytes, recv)| Message::UniStream {
+                    bytes,
+                    src,
+                    recv: RecvStream::new(recv)
+                }),
+                next_bi = next_on_bi_streams(&mut bi_streams, src) =>
+                next_bi.map(|(bytes, send, recv)| Message::BiStream {
+                    bytes,
+                    src,
+                    send: SendStream::new(send),
+                    recv: RecvStream::new(recv)
+                }),
+            };
+            if let Some(message) = message {
+                message_tx
+                    .send((src, message.get_message_data()))
```

Comment:
Why even create the `Message` enum when only the data is used here?

Reply:
Yes. I'm going to remove the Message enum completely.